inst/include/dust/densities.hpp

```diff
@@ -91,11 +90,24 @@ HOSTDEVICE T dnbinom(int x, T size, T mu, bool log) {
   if (mu == 0) {
     return maybe_log(x == 0 ? 0 : -dust::utils::infinity<T>(), log);
   }
-  const T ret = dust::utils::lgamma(static_cast<T>(x + size)) -
-    dust::utils::lgamma(static_cast<T>(size)) -
-    dust::utils::lgamma(static_cast<T>(x + 1)) +
-    size * std::log(prob) + x * std::log(1 - prob);
-  return maybe_log(ret, log);
+
+  // Avoid size / (size + mu) where this would cause prob to be
+  // equal. Somewhat arbitrarily, taking 100 * floating point eps as
+  // the change over.
```

Comment:
```suggestion
  // Avoid size / (size + mu) when size is close to zero, and this would cause prob to be
  // equal to zero. Somewhat arbitrarily, taking 100 * floating point eps as
  // the change over.
```